pkg/ipam/poolallocator/testing/fake_client.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testing
+
+import (
+	"sync"
+
+	"github.com/google/uuid"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	k8stesting "k8s.io/client-go/testing"
+
+	crdv1a2 "antrea.io/antrea/pkg/apis/crd/v1alpha2"
+	fakeversioned "antrea.io/antrea/pkg/client/clientset/versioned/fake"
+)
+
+// Simple client is not sufficient for pool allocator testing,
+// since pool allocator relies on both crd client and pool informer
+// to work in sync. This client extension mimics the real client in
+// conflict handling functionality - pool update will return conflict
+// error unless APIVersion for the updated pool reflect the version
+// stored in the client.
+type IPPoolClientset struct {
+	fakeversioned.Clientset
+	pools   sync.Map
+	watcher *watch.RaceFreeFakeWatcher
+}
+
+func (c *IPPoolClientset) InitPool(pool *crdv1a2.IPPool) {
+	pool.APIVersion = uuid.New().String()
+	c.pools.Store(pool.Name, pool)
+
+	c.watcher.Add(pool)
+}
+
+func NewIPPoolClient() *IPPoolClientset {
+
+	crdClient := &IPPoolClientset{watcher: watch.NewRaceFreeFake()}
+	// map needs to be != nil
+	crdClient.pools.Store("placeholder", nil)
+
+	crdClient.AddReactor("update", "ippools", func(action k8stesting.Action) (bool, runtime.Object, error) {
+		updatedPool := action.(k8stesting.UpdateAction).GetObject().(*crdv1a2.IPPool)
+		obj, exists := crdClient.pools.Load(updatedPool.Name)
+		if !exists {
+			return false, nil, nil
+		}
+		storedPool := obj.(*crdv1a2.IPPool)
+		if storedPool.APIVersion != updatedPool.APIVersion {
+			return true, nil, &errors.StatusError{ErrStatus: metav1.Status{Reason: metav1.StatusReasonConflict, Message: "pool status update conflict"}}
+		}
+
+		updatedPool.APIVersion = uuid.New().String()
```

Comment:
APIVersion is something like `apps/v1`,  "admission.k8s.io/v1", I guess you wanted to use resourceVersion

Reply:
Will change, thank you! Any field will do for the sake of this test, but of course I should use the correct one.

Reply:
Done